subsys/net/ip/net_if.c

```diff
@@ -3296,32 +3689,32 @@ struct in_addr *net_if_ipv4_get_global_addr(struct net_if *iface,
 
 struct net_if *net_if_select_src_iface(const struct sockaddr *dst)
 {
-	struct net_if *iface;
+	struct net_if *iface = NULL;
 
 	if (!dst) {
 		goto out;
 	}
 
+	k_mutex_lock(&lock, K_FOREVER);
+
 	if (IS_ENABLED(CONFIG_NET_IPV6) && dst->sa_family == AF_INET6) {
 		iface = net_if_ipv6_select_src_iface(&net_sin6(dst)->sin6_addr);
-		if (!iface) {
-			goto out;
-		}
-
-		return iface;
+		goto out;
 	}
 
 	if (IS_ENABLED(CONFIG_NET_IPV4) && dst->sa_family == AF_INET) {
 		iface = net_if_ipv4_select_src_iface(&net_sin(dst)->sin_addr);
-		if (!iface) {
-			goto out;
-		}
-
-		return iface;
+		goto out;
 	}
 
 out:
-	return net_if_get_default();
+	k_mutex_unlock(&lock);
+
+	if (iface == NULL) {
+		iface = net_if_get_default();
```

Comment:
The other invocations of `net_if_get_default()` seem to be done with the lock held.

Reply:
> The other invocations of `net_if_get_default()` seem to be done with the lock held.

The `net_if_get_default()` does not really need any locking. It is just accessing net_if array which is created at built time and elements which do not change at runtime.

Reply:
OK, but if that ever changes this could be a problem.  Just looking, what if one of the interfaces' `offload` pointer is changed?  Or if the set of interfaces that are actually available changes at runtime?

I don't insist it be moved into the lock, but it doesn't seem there's a good rationale for *not* having it there.

Reply:
The offload pointer is set once when the offload driver is started (at boot). There can be only one offload driver / interface. There should be no need for locks here.

Reply:
In principle it should to be possible to enable and disable them at runtime, possibly changing the default one.  But if you don't want to move the unlock so it's protected like it is in all the other call sites that's fine.